File: rpn-calculator3.js
// 逆ポーランド記法電卓 (関数型スタイル2)

(function () {
  "use strict";

  var _ = typeof require !== "undefined" ? require("underscore") : this._;

  var OPERATORS = {
    "+": function (x, y) {
      return y + x;
    },
    "-": function (x, y) {
      return y - x;
    },
    "*": function (x, y) {
      return y * x;
    },
    "/": function (x, y) {
      return y / x;
    }
  };

  var calculate = function (rpn) {
    var calc = function (stack, x) {
      if (/^\d+$/.test(x)) {
        return [parseInt(x, 10)].concat(stack);
      } else if (stack.length >= 2) {
        return [opeFn(x)(stack[0], stack[1])].concat(_.rest(stack, 2));
      } else {
        throw new Error("unexpected pattern found");
      }
    };
    return _.first(_.reduce(rpn.split(/\s+/), calc, []));
  };

  var opeFn = function (ope) {
    if (_.has(OPERATORS, ope)) {
      return OPERATORS[ope];
    } else {
      throw new Error("unsupported operator '" + ope + "' is used");
    }
  };

  var RPNCalculator3 =  {
    calculate: calculate
  };
  if (typeof module !== "undefined" && module.exports) {
    module.exports = RPNCalculator3;
  } else {
    this.RPNCalculator3 = RPNCalculator3;
  }

  // 利用例
  var rpn = "1 2 + 3 / 4 - 5 *";
  console.log(rpn, "\n = ", RPNCalculator3.calculate(rpn));
}.call(this));

File: rpn_calculator3.rb
# 逆ポーランド記法電卓 (関数型スタイル2)

module RPNCalculator3
  OPERATORS = {
    "+" => ->(x, y) { y + x },
    "-" => ->(x, y) { y - x },
    "*" => ->(x, y) { y * x },
    "/" => ->(x, y) { y.fdiv(x) }
  }

  module_function

  def calculate(rpn)
    rpn.split.inject([]) { |stack, x|
      if x =~ /^\d+$/
        [x.to_i] + stack
      elsif stack.length >= 2
        [ope_fn(x).(stack[0], stack[1])] + stack[2..-1]
      else
        raise ArgumentError, "unexpected pattern found"
      end
    }[0]
  end

  def ope_fn(ope)
    if OPERATORS.include?(ope)
      OPERATORS[ope]
    else
      raise ArgumentError, "unsupported operator '#{ope}' is used"
    end
  end

  private_class_method :ope_fn
end

if __FILE__ == $0
  # 利用例
  rpn = "1 2 + 3 / 4 - 5 *"
  print rpn, "\n = ", RPNCalculator3.calculate(rpn), "\n"
end
